File: packages/whisker-webview/android/src/main/kotlin/rs/whisker/elements/webview/WhiskerWebView.kt
// Whisker module view hosting a native android.webkit.WebView. Registration
// is driven by `WebViewModule`'s `definition()`, not by annotations on
// this class.
//
// ## Content loading
//
// `url` and `html` are mutually exclusive; a non-empty `url` wins. `url`
// is a one-way-down controlled prop: `lastLoadedUrl` tracks only what the
// prop asked for, so internal navigations (redirects, pushState) are
// never echoed back as a fresh loadUrl().
//
// ## JS bridge
//
// `window.whisker` is injected at document start via
// WebViewCompat.addDocumentStartJavaScript, falling back to onPageStarted
// injection where that feature is unsupported. The shim wires
// `window.whisker.postMessage(data)` up to the JavascriptInterface, and
// `window.whisker._receive(…)` down from `postMessageToPage`.
//
// ## Event dispatch
//
// WebViewClient / WebChromeClient callbacks fire on the UI thread and
// dispatch SYNCHRONOUSLY. They can arrive while the Host is mid-teardown, and
// that is only safe because the Rust renderer is re-entrancy-safe
// (whisker #3: `with_renderer` takes a shared borrow and every renderer
// field borrow is scoped so it never spans a re-entrant FFI call).
//
// The ONE exception is `@JavascriptInterface`, which fires on JavaBridge,
// a background thread. That path must hop to the UI thread via
// `view().post { … }` before touching any Android View / Host event emitter
// state — a genuine thread transition, not a reentrancy guard.
//
// ## Teardown
//
// An OnAttachStateChangeListener tears the web process down on detach so
// the renderer process is released promptly rather than leaking after
// the Host removes the element.

package rs.whisker.elements.webview

import android.annotation.SuppressLint
import android.content.Context
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.webkit.JavascriptInterface
import android.webkit.WebChromeClient
import android.webkit.WebResourceError
import android.webkit.WebResourceRequest
import android.webkit.WebSettings
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.webkit.WebViewCompat
import androidx.webkit.WebViewFeature
import rs.whisker.runtime.WhiskerContext
import rs.whisker.runtime.WhiskerCustomEvent
import rs.whisker.runtime.WhiskerUI

open class WhiskerWebView(context: WhiskerContext) :
    WhiskerUI<android.webkit.WebView>(context) {

    // -------------------------------------------------------------------------
    // JS shim injected at document start
    // -------------------------------------------------------------------------

    companion object {
        /**
         * Injected into every document before any page script runs.
         *
         * - `window.whisker.postMessage(data)` — page → Rust.
         *   Delegates to `__whisker_android.postMessage(s)` (the
         *   JavascriptInterface), serialising non-string args to JSON.
         * - `window.whisker._receive(data)` — Rust → page.
         *   Calls `window.whisker.onMessage` if the page has set it.
         */
        private const val JS_SHIM = """
(function() {
  if (!window.whisker) { window.whisker = {}; }
  window.whisker.postMessage = function(data) {
    var s = (typeof data === 'string') ? data : JSON.stringify(data);
    window.__whisker_android.postMessage(s);
  };
  window.whisker._receive = function(data) {
    if (typeof window.whisker.onMessage === 'function') {
      window.whisker.onMessage(data);
    }
  };
})();
"""
    }

    // -------------------------------------------------------------------------
    // State
    // -------------------------------------------------------------------------

    /** The URL most recently passed to loadUrl(). Re-loading the same URL
     *  would interrupt an in-flight navigation and reset the browser's own
     *  history, so a repeat of this value is skipped. */
    private var lastLoadedUrl: String? = null

    /** The current `html` prop value. Only rendered when `url` is empty. */
    private var pendingHtml: String = ""

    private var scrollEnabled: Boolean = true

    private val cleanupHandler = Handler(Looper.getMainLooper())

    /** Glob patterns; an empty list means "allow all". */
    private var originWhitelist: List<String> = listOf("https://*", "http://*")

    // -------------------------------------------------------------------------
    // View creation
    // -------------------------------------------------------------------------

    @SuppressLint("SetJavaScriptEnabled")
    override fun createView(context: Context): android.webkit.WebView {
        val wv = android.webkit.WebView(context)

        wv.settings.apply {
            javaScriptEnabled = false
            domStorageEnabled = true
            // The component never loads file:// URLs, so denying local-file
            // and content:// access costs nothing and stops a file:// page —
            // or a redirect to one — reading app-sandbox or device files.
            // `allowFileAccess` defaults to true below API 30.
            allowFileAccess = false
            allowContentAccess = false
            // The default whitelist includes http://*, so mixed content has
            // to be permitted for those pages to render.
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                mixedContentMode = WebSettings.MIXED_CONTENT_COMPATIBILITY_MODE
            }
        }

        // Installed before any load, so it is present by the time page
        // scripts run.
        wv.addJavascriptInterface(WhiskerBridge(), "__whisker_android")

        // The onPageStarted fallback injects slightly later than a
        // document-start script, but still before user scripts that wait on
        // DOMContentLoaded.
        val shimInjectedViaCompat = if (WebViewFeature.isFeatureSupported(
                WebViewFeature.DOCUMENT_START_SCRIPT)
        ) {
            WebViewCompat.addDocumentStartJavaScript(wv, JS_SHIM, setOf("*"))
            true
        } else {
            false
        }

        // Inner objects so they can capture `shimInjectedViaCompat` without
        // an extra field.
        wv.webViewClient = object : WebViewClient() {
            override fun onPageStarted(
                view: android.webkit.WebView,
                url: String?,
                favicon: android.graphics.Bitmap?,
            ) {
                if (!shimInjectedViaCompat) {
                    view.evaluateJavascript(JS_SHIM, null)
                }
                val safeUrl = url ?: ""
                WhiskerCustomEvent.dispatch(
                    ui = this@WhiskerWebView,
                    name = "load_start",
                    params = mapOf("url" to safeUrl),
                )
            }

            override fun onPageFinished(view: android.webkit.WebView, url: String?) {
                val safeUrl = url ?: ""
                WhiskerCustomEvent.dispatch(
                    ui = this@WhiskerWebView,
                    name = "load",
                    params = mapOf("url" to safeUrl),
                )
            }

            override fun onReceivedError(
                view: android.webkit.WebView,
                request: WebResourceRequest?,
                error: WebResourceError?,
            ) {
                // Sub-resource errors (images, fonts) would spam the Rust
                // side with non-actionable events.
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M &&
                    request?.isForMainFrame == false) return

                val url = request?.url?.toString() ?: ""
                val code = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    error?.errorCode ?: -1
                } else {
                    -1
                }
                val description = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    error?.description?.toString() ?: ""
                } else {
                    ""
                }
                WhiskerCustomEvent.dispatch(
                    ui = this@WhiskerWebView,
                    name = "error",
                    params = mapOf(
                        "url" to url,
                        "code" to code,
                        "description" to description,
                    ),
                )
            }

            override fun shouldOverrideUrlLoading(
                view: android.webkit.WebView,
                request: WebResourceRequest?,
            ): Boolean {
                val url = request?.url?.toString() ?: return false
                return handleNavigation(view, url)
            }

            @Suppress("OVERRIDE_DEPRECATION")
            override fun shouldOverrideUrlLoading(
                view: android.webkit.WebView,
                url: String?,
            ): Boolean {
                val safeUrl = url ?: return false
                return handleNavigation(view, safeUrl)
            }

            /**
             * Apply the origin whitelist and emit `navigation`. Returns true
             * to block (URL not whitelisted), false to let the load proceed.
             *
             * `navigation` is emitted in BOTH cases, not just when allowed.
             * A custom scheme — an in-app OAuth flow's redirect URI, say —
             * never matches the whitelist and WebView could never load it
             * anyway, so this denial is the only place Rust can observe the
             * attempted URL, and thus the auth code in its query string,
             * without a separate native module.
             */
            private fun handleNavigation(
                view: android.webkit.WebView,
                url: String,
            ): Boolean {
                val allowed = originWhitelist.isEmpty() ||
                    originWhitelist.any { pattern -> matchesGlob(pattern, url) }
                WhiskerCustomEvent.dispatch(
                    ui = this@WhiskerWebView,
                    name = "navigation",
                    params = mapOf("url" to url),
                )
                return !allowed // true = block, false = proceed with load
            }
        }

        wv.webChromeClient = object : WebChromeClient() {
            override fun onProgressChanged(view: android.webkit.WebView, newProgress: Int) {
                val fraction = newProgress / 100.0
                WhiskerCustomEvent.dispatch(
                    ui = this@WhiskerWebView,
                    name = "progress",
                    params = mapOf("progress" to fraction),
                )
            }
        }

        wv.addOnAttachStateChangeListener(
            object : android.view.View.OnAttachStateChangeListener {
                override fun onViewAttachedToWindow(v: android.view.View) {}

                override fun onViewDetachedFromWindow(v: android.view.View) {
                    val webView = v as? android.webkit.WebView ?: return
                    // A Host reparent can detach and attach synchronously in
                    // one frame. Defer destruction so that path keeps the
                    // live browsing context.
                    cleanupHandler.post {
                        if (!webView.isAttachedToWindow) {
                            webView.stopLoading()
                            webView.removeJavascriptInterface("__whisker_android")
                            webView.destroy()
                        }
                    }
                }
            }
        )

        return wv
    }

    // -------------------------------------------------------------------------
    // Props called from WebViewModule
    // -------------------------------------------------------------------------

    /**
     * `url` prop setter. The differs-from-loaded check keeps every Rust
     * re-render that touches an unrelated prop from re-loading the page.
     */
    fun setUrl(incoming: String) {
        val wv = view()
        if (incoming.isEmpty()) {
            if (lastLoadedUrl != null) {
                lastLoadedUrl = null
                loadInlineContent(wv)
            }
            return
        }
        if (incoming == lastLoadedUrl) return
        lastLoadedUrl = incoming
        wv.loadUrl(incoming)
    }

    /**
     * `html` prop setter. When a `url` is active the HTML is stored but not
     * rendered — `url` takes priority.
     */
    fun setHtml(html: String) {
        pendingHtml = html
        val wv = view()
        if (lastLoadedUrl != null && lastLoadedUrl!!.isNotEmpty()) return
        loadInlineContent(wv)
    }

    /** `user-agent` prop setter. Must be set before any load to take effect. */
    fun setUserAgent(ua: String) {
        val wv = view()
        wv.settings.userAgentString = ua.ifEmpty { null }
    }

    @SuppressLint("SetJavaScriptEnabled")
    fun setJavascriptEnabled(enabled: Boolean) {
        val wv = view()
        wv.settings.javaScriptEnabled = enabled
    }

    /**
     * `scroll-enabled` prop setter. Disabling only blocks touch-driven
     * scrolling; programmatic `scrollTo()` still works.
     */
    fun setScrollEnabled(enabled: Boolean) {
        scrollEnabled = enabled
        val wv = view()
        wv.isVerticalScrollBarEnabled = scrollEnabled
        wv.isHorizontalScrollBarEnabled = scrollEnabled
        if (!scrollEnabled) {
            wv.setOnTouchListener { _, event ->
                // Consume only the drag, so ACTION_DOWN / ACTION_UP still
                // reach the page and links keep working.
                event.action == android.view.MotionEvent.ACTION_MOVE
            }
        } else {
            wv.setOnTouchListener(null)
        }
    }

    /**
     * `origin-whitelist` prop setter, taking a JSON array string of glob
     * patterns. A blank value restores the http/https default; malformed
     * JSON keeps the list in force, so a typo can't silently open up or
     * block all navigation.
     */
    fun setOriginWhitelist(json: String) {
        if (json.isBlank()) {
            originWhitelist = listOf("https://*", "http://*")
            return
        }
        try {
            val arr = org.json.JSONArray(json)
            val list = mutableListOf<String>()
            for (i in 0 until arr.length()) {
                list.add(arr.getString(i))
            }
            originWhitelist = list
        } catch (_: Throwable) {
            // Keep the list in force.
        }
    }

    // -------------------------------------------------------------------------
    // Callable UI methods (invoked from WebViewModule's Function blocks)
    // -------------------------------------------------------------------------

    fun reload() {
        view().reload()
    }

    fun goBack() {
        val wv = view()
        if (wv.canGoBack()) wv.goBack()
    }

    fun goForward() {
        val wv = view()
        if (wv.canGoForward()) wv.goForward()
    }

    fun stopLoading() {
        view().stopLoading()
    }

    /**
     * Rust → page message delivery, via `window.whisker._receive(data)`.
     */
    fun postMessageToPage(data: String) {
        val wv = view()
        // Encode as a JSON string literal so the page receives a JS string
        // rather than a bare token that would break the injected expression.
        val encoded = org.json.JSONObject.quote(data)
        wv.evaluateJavascript("window.whisker._receive($encoded)", null)
    }

    /** Run [script] in the page for its side effects. */
    fun evaluateJs(script: String) {
        view().evaluateJavascript(script, null)
    }

    // -------------------------------------------------------------------------
    // Inner JS bridge interface
    // -------------------------------------------------------------------------

    /**
     * Receives `window.whisker.postMessage(data)` calls from the page.
     *
     * `@JavascriptInterface` methods run on the JavaBridge background
     * thread, so the hop back to the UI thread via `view().post { … }` must
     * happen before any View API or Host event emitter state is touched.
     */
    private inner class WhiskerBridge {
        @JavascriptInterface
        fun postMessage(data: String) {
            view().post {
                WhiskerCustomEvent.dispatch(
                    ui = this@WhiskerWebView,
                    name = "message",
                    params = mapOf("data" to data),
                )
            }
        }
    }

    // -------------------------------------------------------------------------
    // Internal helpers
    // -------------------------------------------------------------------------

    private fun loadInlineContent(webView: android.webkit.WebView) {
        if (pendingHtml.isEmpty()) {
            webView.loadUrl("about:blank")
        } else {
            webView.loadDataWithBaseURL(null, pendingHtml, "text/html", "utf-8", null)
        }
    }

    /**
     * Match a URL against a glob pattern. Only `*` is meaningful; every
     * other character is a literal. An empty pattern list is "allow all",
     * handled by the caller.
     */
    private fun matchesGlob(pattern: String, url: String): Boolean {
        val parts = pattern.split("*")
        if (parts.size == 1) return url == pattern // no wildcard: exact match
        var pos = 0
        for ((idx, part) in parts.withIndex()) {
            if (part.isEmpty()) continue
            val found = url.indexOf(part, pos)
            if (found == -1) return false
            // A pattern with no leading `*` must match from the start.
            if (idx == 0 && found != 0) return false
            pos = found + part.length
        }
        // Likewise, no trailing `*` means it must reach the end.
        val lastPart = parts.last()
        if (lastPart.isNotEmpty() && !url.endsWith(lastPart)) return false
        return true
    }
}

File: packages/whisker-webview/ios/Sources/WhiskerWebview/WhiskerWebView.swift
// Lynx UI subclass hosting a WKWebView behind a unified Whisker interface.
// Registration is driven by `WebViewModule`'s `definition()` — no
// annotations required here.
//
// `@objc(WhiskerWebViewView)` pins the Obj-C class name so the codegen
// plugin's `NSClassFromString` lookup can find it regardless of whether the
// SwiftPM-target prefix (`whisker_webview.WhiskerWebViewView`) or the bare
// form is used.
//
// ## WKWebView memory-management
//
// WKWebView retains its `WKUserContentController`, which retains every
// registered `WKScriptMessageHandler`. Registering `WhiskerWebViewView`
// directly would therefore form a retain cycle and the Lynx-owned view
// would never deallocate. `WeakScriptMessageProxy` breaks it: it holds
// `self` weakly, so once the view is gone the bridge callback silently
// drops incoming messages.
//
// ## Event dispatch
//
// Every `WhiskerCustomEvent.dispatch(...)` fires SYNCHRONOUSLY.
// Navigation-delegate / KVO / script-message callbacks can fire during
// Lynx teardown while a renderer op is on the Rust stack, which is only
// safe because the Rust renderer is re-entrancy-safe (whisker #3: shared
// `with_renderer` borrow, `&self` `DynRenderer` methods, FFI-scoped
// per-field `RefCell`s in `BridgeRenderer`).
//
// ## Event payload shape
//
// Params are passed DIRECTLY (e.g. `["url": urlString]`). Do NOT wrap in a
// `detail` key — the iOS bridge's `LynxCustomEvent.params` normalisation
// already places the dispatched params under `detail` in the event body, so
// the Rust structs (`NavEvent { detail: { url } }`, etc.) read the correct
// shape. Double-wrapping produces `detail: { detail: { url } }` and every
// handler receives the default-deserialized empty value.
//
// ## Origin-whitelist glob matching
//
// `*` matches any substring and is the only wildcard, matching the Rust
// contract. Matching is against the full URL string, so `https://*` admits
// any URL whose string starts with `https://`.

import Foundation
import UIKit
import WebKit
import WhiskerModule

// MARK: - Weak proxy (retain-cycle breaker)

/// Forwarding proxy that holds `WhiskerWebViewView` weakly and is
/// registered as the `WKScriptMessageHandler`. When the owning view is
/// deallocated the proxy's weak reference becomes nil and incoming messages
/// are silently dropped.
private final class WeakScriptMessageProxy: NSObject, WKScriptMessageHandler {
    weak var target: WhiskerWebViewView?
    init(target: WhiskerWebViewView) { self.target = target }

    func userContentController(
        _ userContentController: WKUserContentController,
        didReceive message: WKScriptMessage
    ) {
        target?.handleScriptMessage(message)
    }
}

// MARK: - WhiskerWebViewView

@objc(WhiskerWebViewView)
public final class WhiskerWebViewView: WhiskerUI<UIView> {

    // MARK: - Hosted views

    /// Transparent container that fills the LynxUI frame; holds the
    /// `WKWebView` as a subview pinned to its bounds.
    private lazy var containerView: UIView = {
        let v = UIView()
        v.backgroundColor = .clear
        return v
    }()

    /// The live WKWebView. Created once in `createView()`; never replaced.
    private var webView: WKWebView!

    // MARK: - Cached prop state

    private var lastLoadedUrl: String = ""

    /// Cached HTML string. Applied when `url` is empty.
    private var cachedHtml: String = ""

    /// Cached `url` prop. When non-empty, takes precedence over `html`.
    private var cachedUrl: String = ""

    private var originWhitelist: [String] = ["https://*", "http://*"]

    // MARK: - KVO

    /// Retaining this token is what keeps the `estimatedProgress`
    /// observation alive; clearing it cancels the observation.
    private var progressObservation: NSKeyValueObservation?

    // MARK: - LynxUI lifecycle

    @objc public override func createView() -> UIView {
        let config = WKWebViewConfiguration()

        // Document-start injection so page scripts can call
        // `window.whisker.postMessage(data)` immediately.
        let bridgeScript = WKUserScript(
            source: """
            window.whisker = window.whisker || {};
            window.whisker.postMessage = function(data) {
                var s = (typeof data === 'string') ? data : JSON.stringify(data);
                window.webkit.messageHandlers.whisker.postMessage(s);
            };
            window.whisker._receive = function(data) {
                if (window.whisker.onMessage) { window.whisker.onMessage(data); }
            };
            """,
            injectionTime: .atDocumentStart,
            forMainFrameOnly: false
        )
        config.userContentController.addUserScript(bridgeScript)

        // The proxy, not `self` — see the retain-cycle note at the top.
        let proxy = WeakScriptMessageProxy(target: self)
        config.userContentController.add(proxy, name: "whisker")

        if #available(iOS 14.0, *) {
            config.defaultWebpagePreferences.allowsContentJavaScript = true
        }

        let wv = WKWebView(frame: .zero, configuration: config)
        wv.navigationDelegate = self
        self.webView = wv

        // The block fires on the main queue.
        progressObservation = wv.observe(
            \.estimatedProgress,
            options: [.new]
        ) { [weak self] _, change in
            guard let self, let progress = change.newValue else { return }
            self.emitProgress(progress)
        }

        containerView.addSubview(wv)

        // `setUrl` / `setHtml` optional-chain on `webView` and silently
        // no-op when props arrive before `createView()` runs, so replay
        // whatever was cached.
        if !cachedUrl.isEmpty {
            lastLoadedUrl = cachedUrl
            if let url = URL(string: cachedUrl) {
                wv.load(URLRequest(url: url))
            }
        } else if !cachedHtml.isEmpty {
            wv.loadHTMLString(cachedHtml, baseURL: nil)
        }

        return containerView
    }

    @objc public override func frameDidChange() {
        super.frameDidChange()
        webView?.frame = self.view().bounds
    }

    /// `WhiskerUI` / `LynxUI` exposes no teardown override, so cleanup has
    /// to happen in `deinit`. That works because nothing retains this
    /// object back: `WeakScriptMessageProxy` stands in for us on the
    /// `WKUserContentController` and `navigationDelegate` is weak, so
    /// `deinit` does fire — on the main thread — when Lynx releases the
    /// view, and the web process is freed promptly.
    deinit {
        progressObservation = nil
        webView?.stopLoading()
        webView?.navigationDelegate = nil
        webView?.configuration.userContentController.removeScriptMessageHandler(forName: "whisker")
        webView?.configuration.userContentController.removeAllUserScripts()
    }

    // MARK: - Public setters (called by WebViewModule's Prop closures)

    // ---- Content ---------------------------------------------------------

    /// Set the `url` prop. The change-detection guard stops reactive
    /// re-renders that don't touch the value from reloading the page.
    public func setUrl(_ urlString: String) {
        cachedUrl = urlString
        guard !urlString.isEmpty else { return }
        guard urlString != lastLoadedUrl else { return }
        lastLoadedUrl = urlString
        guard let url = URL(string: urlString) else { return }
        webView?.load(URLRequest(url: url))
    }

    /// Set the `html` prop. Applied only when `url` is empty so `url` takes
    /// precedence.
    public func setHtml(_ html: String) {
        cachedHtml = html
        guard cachedUrl.isEmpty else { return }
        webView?.loadHTMLString(html, baseURL: nil)
    }

    // ---- Browser behaviour -----------------------------------------------

    public func setUserAgent(_ ua: String) {
        webView?.customUserAgent = ua.isEmpty ? nil : ua
    }

    public func setJavascriptEnabled(_ s: String) {
        guard #available(iOS 14.0, *) else { return }
        webView?.configuration.defaultWebpagePreferences.allowsContentJavaScript = (s != "false")
    }

    public func setScrollEnabled(_ s: String) {
        webView?.scrollView.isScrollEnabled = (s != "false")
    }

    /// Parses a JSON-array string like `["https://*","http://*"]` into the
    /// local `originWhitelist` used by `decidePolicyFor`.
    public func setOriginWhitelist(_ json: String) {
        guard !json.isEmpty else { return }
        // Hand-rolled quoted-token scan rather than JSONSerialization: the
        // only legal input is a JSON string array from the Rust
        // `origin_whitelist_json` helper.
        var patterns: [String] = []
        var idx = json.startIndex
        while idx < json.endIndex {
            guard let open = json[idx...].firstIndex(of: "\"") else { break }
            let afterOpen = json.index(after: open)
            guard afterOpen < json.endIndex else { break }
            // Scan to the closing quote, honouring `\"` escapes.
            var end = afterOpen
            while end < json.endIndex {
                if json[end] == "\\" {
                    let next = json.index(after: end)
                    if next < json.endIndex { end = json.index(after: next) } else { end = next }
                } else if json[end] == "\"" {
                    break
                } else {
                    end = json.index(after: end)
                }
            }
            let raw = String(json[afterOpen..<end])
            let unescaped = raw
                .replacingOccurrences(of: "\\\"", with: "\"")
                .replacingOccurrences(of: "\\\\", with: "\\")
            patterns.append(unescaped)
            idx = end < json.endIndex ? json.index(after: end) : json.endIndex
        }
        if !patterns.isEmpty {
            originWhitelist = patterns
        }
    }

    // MARK: - Imperative method targets (called by WebViewModule's Function closures)

    public func reloadPage() {
        webView?.reload()
    }

    public func goBackPage() {
        webView?.goBack()
    }

    public func goForwardPage() {
        webView?.goForward()
    }

    public func stopLoadingPage() {
        webView?.stopLoading()
    }

    /// Deliver a Rust-originated string to the page's `window.whisker.onMessage`
    /// handler by evaluating `window.whisker._receive(...)` in the page context.
    public func postMessageToPage(_ data: String) {
        // Encoded as a JS string literal so embedded quotes, backslashes,
        // and newlines can't break the injected script.
        let jsString = jsonStringLiteral(data)
        webView?.evaluateJavaScript("window.whisker._receive(\(jsString))", completionHandler: nil)
    }

    /// Evaluate arbitrary JavaScript for its side effects. The returned
    /// `value` is always the empty string — the completion fires after
    /// this synchronous dispatch returns; use
    /// [`evaluateJavaScript(_:resolving:)`] for the result.
    public func evaluateJavaScript(_ script: String) -> WhiskerValue {
        guard let wv = webView else {
            return .map(["value": .string("")])
        }
        wv.evaluateJavaScript(script, completionHandler: nil)
        return .map(["value": .string("")])
    }

    /// Evaluate JavaScript and settle `promise` from the completion:
    /// the result as a JSON-encoded string (`"null"` when the script
    /// yields no value), matching Android's `evaluateJavascript`
    /// callback convention; a JS exception rejects.
    public func evaluateJavaScript(_ script: String, resolving promise: WhiskerPromise) {
        guard let wv = webView else {
            promise.resolve(.string("null"))
            return
        }
        wv.evaluateJavaScript(script) { value, error in
            if let error = error {
                // The JS exception text rides in userInfo, not
                // localizedDescription.
                let ns = error as NSError
                let detail = ns.userInfo["WKJavaScriptExceptionMessage"] as? String
                    ?? error.localizedDescription
                promise.reject("evaluateJavaScript failed: \(detail)")
                return
            }
            guard let value = value else {
                promise.resolve(.string("null"))
                return
            }
            // The precheck matters: `JSONSerialization.data` raises an
            // (uncatchable) NSException for non-JSON top-level objects.
            guard JSONSerialization.isValidJSONObject(value)
                || value is NSString || value is NSNumber || value is NSNull,
                let data = try? JSONSerialization.data(
                    withJSONObject: value, options: .fragmentsAllowed),
                let json = String(data: data, encoding: .utf8)
            else {
                promise.reject("evaluateJavaScript: result is not JSON-encodable")
                return
            }
            promise.resolve(.string(json))
        }
    }

    public func canGoBackResult() -> WhiskerValue {
        return .bool(webView?.canGoBack ?? false)
    }

    public func canGoForwardResult() -> WhiskerValue {
        return .bool(webView?.canGoForward ?? false)
    }

    // MARK: - Script-message handler (called by the proxy)

    /// Called by `WeakScriptMessageProxy` when the page invokes
    /// `window.whisker.postMessage(...)`.
    func handleScriptMessage(_ message: WKScriptMessage) {
        let data = message.body as? String ?? ""
        emitMessage(data)
    }

    // MARK: - Event emission helpers

    // These dispatch SYNCHRONOUSLY, which is only safe because the Rust
    // renderer is re-entrancy-safe: `DynRenderer` methods take `&self`,
    // `BridgeRenderer` keeps its state behind per-field `RefCell`s with
    // FFI-scoped borrows, and `with_renderer` takes a SHARED borrow
    // (whisker #3). Navigation-delegate / KVO / script-message callbacks can
    // fire during Lynx teardown while `remove_child` is on the Rust stack,
    // so a re-entrant dispatch is granted rather than aborting. Deferring a
    // runloop tick instead would cost every webview event a tick of latency.

    private func emitLoadStart(_ urlString: String) {
        WhiskerCustomEvent.dispatch(from: self, name: "load_start", params: ["url": urlString])
    }

    private func emitLoad(_ urlString: String) {
        WhiskerCustomEvent.dispatch(from: self, name: "load", params: ["url": urlString])
    }

    private func emitError(urlString: String, code: Int, description: String) {
        WhiskerCustomEvent.dispatch(from: self, name: "error", params: [
            "url": urlString,
            "code": code,
            "description": description,
        ])
    }

    private func emitNavigation(_ urlString: String) {
        WhiskerCustomEvent.dispatch(from: self, name: "navigation", params: ["url": urlString])
    }

    private func emitProgress(_ progress: Double) {
        WhiskerCustomEvent.dispatch(from: self, name: "progress", params: ["progress": progress])
    }

    private func emitMessage(_ data: String) {
        WhiskerCustomEvent.dispatch(from: self, name: "message", params: ["data": data])
    }

    // MARK: - Origin-whitelist matching

    /// Returns `true` if `urlString` is allowed by at least one pattern in
    /// `originWhitelist`.
    private func isAllowed(_ urlString: String) -> Bool {
        for pattern in originWhitelist {
            if globMatch(pattern: pattern, string: urlString) { return true }
        }
        return false
    }

    /// Shell-style glob match: `*` matches any substring (including empty).
    /// No `?` wildcard — the Rust contract documents `*` only.
    private func globMatch(pattern: String, string: String) -> Bool {
        let parts = pattern.components(separatedBy: "*")
        guard !parts.isEmpty else { return true }

        var remaining = string[...]

        for (i, part) in parts.enumerated() {
            if part.isEmpty { continue }
            if i == 0 {
                guard remaining.hasPrefix(part) else { return false }
                remaining = remaining.dropFirst(part.count)
            } else if i == parts.count - 1 {
                guard remaining.hasSuffix(part) else { return false }
            } else {
                guard let range = remaining.range(of: part) else { return false }
                remaining = remaining[range.upperBound...]
            }
        }
        return true
    }

    // MARK: - JS string helpers

    /// JSON-encodes a Swift `String` into a JavaScript string literal
    /// (including surrounding double quotes) safe to embed directly in
    /// a `<script>` call. Escapes `"`, `\`, and control characters.
    private func jsonStringLiteral(_ s: String) -> String {
        var out = "\""
        for ch in s.unicodeScalars {
            switch ch {
            case "\"": out += "\\\""
            case "\\": out += "\\\\"
            case "\n": out += "\\n"
            case "\r": out += "\\r"
            case "\t": out += "\\t"
            default:
                if ch.value < 0x20 {
                    out += String(format: "\\u%04X", ch.value)
                } else {
                    out += String(ch)
                }
            }
        }
        out += "\""
        return out
    }
}

// MARK: - WKNavigationDelegate

extension WhiskerWebViewView: WKNavigationDelegate {

    public func webView(
        _ webView: WKWebView,
        decidePolicyFor navigationAction: WKNavigationAction,
        decisionHandler: @escaping (WKNavigationActionPolicy) -> Void
    ) {
        // Sub-resource loads (images, XHR) are always allowed; only
        // main-frame navigation is gated.
        guard navigationAction.targetFrame?.isMainFrame == true else {
            decisionHandler(.allow)
            return
        }

        let urlString = navigationAction.request.url?.absoluteString ?? ""
        let scheme = navigationAction.request.url?.scheme?.lowercased() ?? ""

        // Only real web navigations are gated by the origin whitelist.
        if scheme == "http" || scheme == "https" {
            if !isAllowed(urlString) {
                // Denied, but still observable — see the note on the
                // "everything else" branch below.
                emitNavigation(urlString)
                decisionHandler(.cancel)
                return
            }
            emitNavigation(urlString)
            decisionHandler(.allow)
            return
        }

        // Inline / generated content must bypass the whitelist:
        // `loadHTMLString(_:baseURL:)` navigates to `about:blank`, and
        // `data:` / `blob:` back inline documents. None of them match a
        // `https://*` / `http://*` pattern, so gating them would cancel the
        // load and render a blank page.
        if scheme == "about" || scheme == "data" || scheme == "blob" {
            decisionHandler(.allow)
            return
        }

        // Everything else fails closed — notably `file:`, a local-file
        // disclosure risk, plus `javascript:` and custom deep-link schemes.
        // The component exposes no file-access prop, so no legitimate
        // in-webview navigation targets them.
        //
        // `navigation` is still emitted before cancelling: a custom scheme
        // is exactly how an in-app OAuth flow's redirect URI surfaces, and
        // since WKWebView can never load it, this denial is the only place
        // Rust can observe the attempted URL — and thus the auth code in its
        // query string — without a separate native module.
        emitNavigation(urlString)
        decisionHandler(.cancel)
    }

    public func webView(
        _ webView: WKWebView,
        didStartProvisionalNavigation navigation: WKNavigation!
    ) {
        let urlString = webView.url?.absoluteString ?? ""
        emitLoadStart(urlString)
    }

    public func webView(
        _ webView: WKWebView,
        didFinish navigation: WKNavigation!
    ) {
        let urlString = webView.url?.absoluteString ?? ""
        emitLoad(urlString)
    }

    public func webView(
        _ webView: WKWebView,
        didFail navigation: WKNavigation!,
        withError error: Error
    ) {
        let urlString = webView.url?.absoluteString ?? ""
        let nsErr = error as NSError
        emitError(urlString: urlString, code: nsErr.code, description: nsErr.localizedDescription)
    }

    public func webView(
        _ webView: WKWebView,
        didFailProvisionalNavigation navigation: WKNavigation!,
        withError error: Error
    ) {
        // `didFailProvisionalNavigation` fires before a committed URL is
        // available, so fall back to the request URL stored on the webView.
        let urlString = webView.url?.absoluteString
            ?? webView.backForwardList.currentItem?.url.absoluteString
            ?? ""
        let nsErr = error as NSError
        emitError(urlString: urlString, code: nsErr.code, description: nsErr.localizedDescription)
    }
}
